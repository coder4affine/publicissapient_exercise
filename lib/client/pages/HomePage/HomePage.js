"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactHelmet = require("react-helmet");

var _ResponsiveTable = _interopRequireDefault(require("../../components/ResponsiveTable/ResponsiveTable"));

var _Divider = _interopRequireDefault(require("../../components/Divider/Divider"));

var _RechartLine = _interopRequireDefault(require("../../components/RechartLine"));

var _useNewsList = _interopRequireDefault(require("./useNewsList"));

require("./HomePage.css");

const cols = {
  comments: {
    title: 'Comments',
    style: {
      width: '5%'
    }
  },
  voteCount: {
    title: 'Vote Count',
    style: {
      width: '5%'
    }
  },
  upVote: {
    title: 'UpVote',
    style: {
      width: '5%'
    }
  },
  newsDetails: {
    title: 'News Details',
    style: {
      width: '85%',
      textAlign: 'left'
    },
    rowStyle: {
      textAlign: 'left'
    }
  }
};

const HomePage = ({
  news,
  loading,
  loadNews,
  error,
  upVote,
  hideHit,
  match: {
    params: {
      page
    }
  },
  history: {
    push
  }
}) => {
  const {
    newsList
  } = (0, _useNewsList.default)({
    news,
    upVote,
    hideHit
  });
  (0, _react.useEffect)(() => {
    window.scrollTo(0, 0);
    loadNews(page || 1);
  }, [loadNews, page]);

  const previous = () => {
    if (page && page !== 1) {
      const prevPage = Number(page) - 1;
      push(`/${prevPage === 1 ? '' : prevPage}`);
    }
  };

  const next = () => {
    push(`/${Number(page || 1) + 1}`);
  };

  const head = () => {
    return /*#__PURE__*/_react.default.createElement(_reactHelmet.Helmet, {
      key: Math.random()
    }, /*#__PURE__*/_react.default.createElement("title", null, "Search Heacker News"), /*#__PURE__*/_react.default.createElement("meta", {
      property: "og:title",
      content: "Search Heacker News"
    }), /*#__PURE__*/_react.default.createElement("meta", {
      name: "description",
      content: "Hacker News Search, millions articles and comments at your fingertips."
    }), /*#__PURE__*/_react.default.createElement("meta", {
      name: "robots",
      content: "Hacker, News"
    }), /*#__PURE__*/_react.default.createElement("link", {
      rel: "canonical",
      href: "https://hn.algolia.com/"
    }));
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, head(), loading ? /*#__PURE__*/_react.default.createElement("h2", {
    "data-testid": "h2loading"
  }, "Loading....") : error ? /*#__PURE__*/_react.default.createElement("h2", {
    "data-testid": "h2error"
  }, error) : news.hits && news.hits.length <= 0 ? /*#__PURE__*/_react.default.createElement("h2", {
    "data-testid": "norecords"
  }, "No Records Available") : /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: '#F5F5EF'
    },
    "data-testid": "container"
  }, /*#__PURE__*/_react.default.createElement(_ResponsiveTable.default, {
    columns: cols,
    rows: newsList,
    "data-testid": "responsiveable"
  }), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      display: 'flex',
      flexDirection: 'row',
      justifyContent: 'flex-end',
      color: '#bf4808',
      fontWeight: 'bold',
      fontSize: '1rem',
      padding: 10
    }
  }, page && page !== 1 ? /*#__PURE__*/_react.default.createElement("span", {
    className: "btn",
    "aria-label": "Previous",
    role: "button",
    tabIndex: 0,
    onKeyDown: previous,
    onClick: previous,
    "data-testid": "previousbtn"
  }, "Previous") : null, page && page > 1 && page < news.nbPages - 1 ? /*#__PURE__*/_react.default.createElement("span", {
    style: {
      margin: '0 5px'
    }
  }, "|") : null, !page || page < news.nbPages - 1 ? /*#__PURE__*/_react.default.createElement("span", {
    className: "btn",
    "aria-label": "Next",
    role: "button",
    tabIndex: 0,
    onKeyDown: next,
    onClick: next,
    "data-testid": "nextbtn"
  }, "Next") : null), /*#__PURE__*/_react.default.createElement(_Divider.default, null), newsList.length > 0 ? [/*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: 10
    },
    key: "0"
  }, /*#__PURE__*/_react.default.createElement(_RechartLine.default, {
    newsList: newsList,
    xDataKey: "objectID",
    yDataKey: "voteCount"
  })), /*#__PURE__*/_react.default.createElement(_Divider.default, {
    key: "1"
  })] : null));
};

HomePage.propTypes = {
  news: _propTypes.default.shape({
    hits: _propTypes.default.array,
    page: _propTypes.default.number,
    nbPages: _propTypes.default.number
  }).isRequired,
  loadNews: _propTypes.default.func.isRequired,
  loading: _propTypes.default.bool.isRequired,
  error: _propTypes.default.string,
  match: _propTypes.default.shape({
    params: _propTypes.default.object
  }).isRequired,
  history: _propTypes.default.shape({
    push: _propTypes.default.func.isRequired
  }).isRequired,
  upVote: _propTypes.default.func.isRequired,
  hideHit: _propTypes.default.func.isRequired
};
HomePage.defaultProps = {
  error: ''
};
var _default = HomePage;
exports.default = _default;