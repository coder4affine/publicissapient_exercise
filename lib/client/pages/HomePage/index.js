"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _reactRedux = require("react-redux");

var _HomePage = _interopRequireDefault(require("./HomePage"));

var _actionTypes = require("../../../constants/actionTypes");

var _utils = require("../../../utils");

const mapStateToProps = state => {
  return {
    news: state.news,
    loading: !!state.loading[_actionTypes.FETCH_NEWS],
    error: state.erros[_actionTypes.FETCH_NEWS]
  };
};

const mapDispatchToProps = dispatch => {
  return {
    loadNews: payload => dispatch((0, _utils.action)(`${_actionTypes.FETCH_NEWS}_${_actionTypes.REQUEST}`, payload)),
    upVote: payload => dispatch((0, _utils.action)(`${_actionTypes.UP_VOTE}_${_actionTypes.REQUEST}`, payload)),
    hideHit: payload => dispatch((0, _utils.action)(`${_actionTypes.HIDE_HIT}_${_actionTypes.REQUEST}`, payload))
  };
};

var _default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(_HomePage.default);

exports.default = _default;