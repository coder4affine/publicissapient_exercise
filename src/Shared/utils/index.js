export const action = (type, payload, meta) => ({
  type,
  payload,
  meta,
});

export const a = 'a';
