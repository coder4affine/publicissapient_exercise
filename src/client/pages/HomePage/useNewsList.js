import React, { useState, useEffect } from 'react';
import NewsDetails from '../../components/NewsDetails/NewsDetails';
import UpIcon from '../../assets/icons/arrow_drop_up.svg';

const useNewsList = ({ news, hideHit, upVote }) => {
  const [newsList, setNewsList] = useState([]);

  useEffect(() => {
    const data =
      news?.hits?.reduce((p, c) => {
        if (!c.hide) {
          return [
            ...p,
            {
              comments: c.num_comments || 0,
              voteCount: c.points || 0,
              upVote: (
                <div
                  className="btn"
                  aria-label={`upVote for ${c.objectID}`}
                  role="button"
                  tabIndex={0}
                  data-testid={`upvote-${c.objectID}`}
                  onKeyDown={() => upVote(c.objectID)}
                  onClick={() => upVote(c.objectID)}
                >
                  <UpIcon width={32} height={32} fill="#979797" />
                </div>
              ),
              newsDetails: (
                <NewsDetails hit={c} onHide={hideHit} data-testid={`newsDetails-${c.objectID}`} />
              ),
              objectID: c.objectID,
            },
          ];
        }
        return p;
      }, []) || [];
    setNewsList(data);
  }, [news, hideHit, upVote]);

  return {
    newsList,
  };
};

export default useNewsList;
