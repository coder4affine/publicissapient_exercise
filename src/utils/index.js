export const action = (type, payload, meta) => ({
  type,
  payload,
  meta,
});

export const getItem = (key) => {
  const data = localStorage.getItem(key);
  return JSON.parse(data);
};

export const setItem = (key, value) => {
  localStorage.setItem(key, JSON.stringify(value));
};
