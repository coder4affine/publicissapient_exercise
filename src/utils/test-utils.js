import React from 'react';
import PropTypes from 'prop-types';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from '../redux/store/createStore';
import sagas from '../redux/Sagas';

const reducerInitialState = window.__PRELOADED_STATE__;
delete window.__PRELOADED_STATE__;

const store = configureStore(reducerInitialState);

store.runSaga(sagas);

function render(ui, { ...renderOptions } = {}) {
  const Wrapper = ({ children }) => {
    return <Provider store={store}>{children}</Provider>;
  };
  Wrapper.propTypes = {
    children: PropTypes.element.isRequired,
  };
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
