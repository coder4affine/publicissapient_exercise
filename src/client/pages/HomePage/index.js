import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import loadable from '@loadable/component';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import {} from 'recharts';
import { FETCH_NEWS, REQUEST, UP_VOTE, HIDE_HIT } from '../../../constants/actionTypes';
import { action } from '../../../utils';
import ResponsiveTable from '../../components/ResponsiveTable/ResponsiveTable';
import NewsDetails from '../../components/NewsDetails/NewsDetails';
import Divider from '../../components/Divider/Divider';

import { ReactComponent as UpIcon } from '../../assets/icons/arrow_drop_up.svg';

const Rechart = loadable.lib(() => import(/* webpackChunkName: 'recharts' */ 'recharts'));

const cols = {
  comments: { title: 'Comments', style: { width: '5%' } },
  voteCount: { title: 'Vote Count', style: { width: '5%' } },
  upVote: { title: 'UpVote', style: { width: '5%' } },
  newsDetails: {
    title: 'News Details',
    style: { width: '85%', textAlign: 'left' },
    rowStyle: { textAlign: 'left' },
  },
};

const HomePage = ({
  news,
  loading,
  loadNews,
  error,
  upVote,
  hideHit,
  match: {
    params: { page },
  },
  history: { push },
}) => {
  const [newsList, setNewsList] = useState([]);

  useEffect(() => {
    window.scrollTo(0, 0);
    loadNews(page || 1);
  }, [loadNews, page]);

  useEffect(() => {
    const data =
      news?.hits?.reduce((p, c) => {
        if (!c.hide) {
          return [
            ...p,
            {
              comments: c.num_comments || 0,
              voteCount: c.points || 0,
              upVote: (
                <div
                  className="btn"
                  aria-label={`upVote for ${c.objectID}`}
                  role="button"
                  tabIndex={0}
                  data-testid={`upvote-${c.objectID}`}
                  onKeyDown={() => upVote(c.objectID)}
                  onClick={() => upVote(c.objectID)}
                >
                  <UpIcon width={32} height={32} fill="#979797" />
                </div>
              ),
              newsDetails: (
                <NewsDetails hit={c} onHide={hideHit} data-testid={`newsDetails-${c.objectID}`} />
              ),
              objectID: c.objectID,
            },
          ];
        }
        return p;
      }, []) || [];
    setNewsList(data);
  }, [news, hideHit, upVote]);

  const previous = () => {
    if (page && page !== 1) {
      const prevPage = Number(page) - 1;
      push(`/${prevPage === 1 ? '' : prevPage}`);
    }
  };

  const next = () => {
    push(`/${Number(page || 1) + 1}`);
  };

  const head = () => {
    return (
      <Helmet key={Math.random()}>
        <title>Search Heacker News</title>
        <meta property="og:title" content="Search Heacker News" />
        <meta
          name="description"
          content="Hacker News Search, millions articles and comments at your fingertips."
        />
        <meta name="robots" content="Hacker, News" />
        <link rel="canonical" href="https://hn.algolia.com/" />
      </Helmet>
    );
  };

  return (
    <>
      {head()}
      <Choose>
        <When condition={loading}>
          <h2 data-testid="h2loading">Loading....</h2>
        </When>
        <When condition={error}>
          <h2 data-testid="h2error">{error}</h2>
        </When>
        <When condition={news.hits && news.hits.length <= 0}>
          <h2 data-testid="norecords">No Records Available</h2>
        </When>
        <Otherwise>
          <div
            style={{
              background: '#F5F5EF',
            }}
            data-testid="container"
          >
            <ResponsiveTable columns={cols} rows={newsList} data-testid="responsivetable" />
            <div
              style={{
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'flex-end',
                color: '#bf4808',
                fontWeight: 'bold',
                fontSize: '1rem',
                padding: 10,
              }}
            >
              <If condition={page && page !== 1}>
                <span
                  className="btn"
                  aria-label="Previous"
                  role="button"
                  tabIndex={0}
                  onKeyDown={previous}
                  onClick={previous}
                  data-testid="previousbtn"
                >
                  Previous
                </span>
              </If>
              <If condition={page && page > 1 && page < news.nbPages - 1}>
                <span style={{ margin: '0 5px' }}>|</span>
              </If>
              <If condition={!page || page < news.nbPages - 1}>
                <span
                  className="btn"
                  aria-label="Next"
                  role="button"
                  tabIndex={0}
                  onKeyDown={next}
                  onClick={next}
                  data-testid="nextbtn"
                >
                  Next
                </span>
              </If>
            </div>
            <Divider />
            <If condition={newsList.length > 0}>
              <div style={{ margin: 10 }}>
                <Rechart>
                  {({
                    LineChart,
                    CartesianGrid,
                    XAxis,
                    YAxis,
                    Tooltip,
                    Line,
                    Label,
                    ResponsiveContainer,
                  }) => (
                    <ResponsiveContainer width="100%" height={300}>
                      <LineChart
                        data={newsList}
                        margin={{ top: 10, right: 10, left: 20, bottom: 100 }}
                      >
                        <CartesianGrid vertical={false} />
                        <XAxis
                          dataKey="objectID"
                          angle={-90}
                          textAnchor="end"
                          verticalAnchor="middle"
                        >
                          <Label value="ID" position="bottom" offset={50} />
                        </XAxis>
                        <YAxis label={{ value: 'Votes', angle: -90, position: 'left' }} />
                        <Tooltip />
                        <Line
                          dataKey="voteCount"
                          stroke="#366396"
                          strokeWidth={4}
                          dot={{ strokeWidth: 6, r: 3 }}
                        />
                      </LineChart>
                    </ResponsiveContainer>
                  )}
                </Rechart>
              </div>
              <Divider />
            </If>
          </div>
        </Otherwise>
      </Choose>
    </>
  );
};

HomePage.propTypes = {
  news: PropTypes.shape({
    hits: PropTypes.array,
    page: PropTypes.number,
    nbPages: PropTypes.number,
  }).isRequired,
  loadNews: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  match: PropTypes.shape({
    params: PropTypes.object,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  upVote: PropTypes.func.isRequired,
  hideHit: PropTypes.func.isRequired,
};

HomePage.defaultProps = {
  error: '',
};

const mapStateToProps = (state) => {
  return {
    news: state.news,
    loading: !!state.loading[FETCH_NEWS],
    error: state.erros[FETCH_NEWS],
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    loadNews: (payload) => dispatch(action(`${FETCH_NEWS}_${REQUEST}`, payload)),
    upVote: (payload) => dispatch(action(`${UP_VOTE}_${REQUEST}`, payload)),
    hideHit: (payload) => dispatch(action(`${HIDE_HIT}_${REQUEST}`, payload)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
