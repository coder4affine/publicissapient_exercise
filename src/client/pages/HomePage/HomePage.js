import React, { useEffect } from 'react';
import PropTypes from 'prop-types';

import { Helmet } from 'react-helmet';
import ResponsiveTable from '../../components/ResponsiveTable/ResponsiveTable';
import Divider from '../../components/Divider/Divider';
import RechartLine from '../../components/RechartLine';
import useNewsList from './useNewsList';

import './HomePage.css';

const cols = {
  comments: { title: 'Comments', style: { width: '5%' } },
  voteCount: { title: 'Vote Count', style: { width: '5%' } },
  upVote: { title: 'UpVote', style: { width: '5%' } },
  newsDetails: {
    title: 'News Details',
    style: { width: '85%', textAlign: 'left' },
    rowStyle: { textAlign: 'left' },
  },
};

const HomePage = ({
  news,
  loading,
  loadNews,
  error,
  upVote,
  hideHit,
  match: {
    params: { page },
  },
  history: { push },
}) => {
  const { newsList } = useNewsList({ news, upVote, hideHit });

  useEffect(() => {
    window.scrollTo(0, 0);
    loadNews(page || 1);
  }, [loadNews, page]);

  const previous = () => {
    if (page && page !== 1) {
      const prevPage = Number(page) - 1;
      push(`/${prevPage === 1 ? '' : prevPage}`);
    }
  };

  const next = () => {
    push(`/${Number(page || 1) + 1}`);
  };

  const head = () => {
    return (
      <Helmet key={Math.random()}>
        <title>Search Heacker News</title>
        <meta property="og:title" content="Search Heacker News" />
        <meta
          name="description"
          content="Hacker News Search, millions articles and comments at your fingertips."
        />
        <meta name="robots" content="Hacker, News" />
        <link rel="canonical" href="https://hn.algolia.com/" />
      </Helmet>
    );
  };

  return (
    <>
      {head()}
      <Choose>
        <When condition={loading}>
          <h2 data-testid="h2loading">Loading....</h2>
        </When>
        <When condition={error}>
          <h2 data-testid="h2error">{error}</h2>
        </When>
        <When condition={news.hits && news.hits.length <= 0}>
          <h2 data-testid="norecords">No Records Available</h2>
        </When>
        <Otherwise>
          <div
            style={{
              background: '#F5F5EF',
            }}
            data-testid="container"
          >
            <ResponsiveTable columns={cols} rows={newsList} data-testid="responsiveable" />
            <div
              style={{
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'flex-end',
                color: '#bf4808',
                fontWeight: 'bold',
                fontSize: '1rem',
                padding: 10,
              }}
            >
              <If condition={page && page !== 1}>
                <span
                  className="btn"
                  aria-label="Previous"
                  role="button"
                  tabIndex={0}
                  onKeyDown={previous}
                  onClick={previous}
                  data-testid="previousbtn"
                >
                  Previous
                </span>
              </If>
              <If condition={page && page > 1 && page < news.nbPages - 1}>
                <span style={{ margin: '0 5px' }}>|</span>
              </If>
              <If condition={!page || page < news.nbPages - 1}>
                <span
                  className="btn"
                  aria-label="Next"
                  role="button"
                  tabIndex={0}
                  onKeyDown={next}
                  onClick={next}
                  data-testid="nextbtn"
                >
                  Next
                </span>
              </If>
            </div>
            <Divider />
            <If condition={newsList.length > 0}>
              <div style={{ margin: 10 }}>
                <RechartLine newsList={newsList} xDataKey="objectID" yDataKey="voteCount" />
              </div>
              <Divider />
            </If>
          </div>
        </Otherwise>
      </Choose>
    </>
  );
};

HomePage.propTypes = {
  news: PropTypes.shape({
    hits: PropTypes.array,
    page: PropTypes.number,
    nbPages: PropTypes.number,
  }).isRequired,
  loadNews: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  match: PropTypes.shape({
    params: PropTypes.object,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  upVote: PropTypes.func.isRequired,
  hideHit: PropTypes.func.isRequired,
};

HomePage.defaultProps = {
  error: '',
};

export default HomePage;
