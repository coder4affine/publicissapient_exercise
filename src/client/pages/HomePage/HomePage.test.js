import React from 'react';
import { waitFor, fireEvent } from '@testing-library/react';
import { Helmet } from 'react-helmet';
import axios from 'axios';
import { render, screen } from '../../../utils/test-utils';
import HomePage from './index';

jest.mock('axios');

const data = {
  hits: [
    {
      created_at: '2020-02-12T17:38:20.000Z',
      title: 'Gears',
      url: 'https://ciechanow.ski/gears/',
      author: 'robert-boehnke',
      points: 2760,
      story_text: null,
      comment_text: null,
      num_comments: 222,
      story_id: null,
      story_title: null,
      story_url: null,
      parent_id: null,
      created_at_i: 1581529100,
      _tags: ['story', 'author_robert-boehnke', 'story_22310813'],
      objectID: '22310813',
      _highlightResult: {
        title: {
          value: 'Gears',
          matchLevel: 'none',
          matchedWords: [],
        },
        url: {
          value: 'https://ciechanow.ski/gears/',
          matchLevel: 'none',
          matchedWords: [],
        },
        author: {
          value: 'robert-boehnke',
          matchLevel: 'none',
          matchedWords: [],
        },
      },
    },
    {
      created_at: '2013-04-12T05:07:53.000Z',
      title: '',
      url: '',
      author: 'xijuan',
      points: 2751,
      story_text: '26 to 30',
      comment_text: null,
      num_comments: null,
      story_id: null,
      story_title: null,
      story_url: null,
      parent_id: 5536734,
      created_at_i: 1365743273,
      relevancy_score: 4557,
      _tags: ['pollopt', 'author_xijuan', 'story_5536739'],
      objectID: '5536739',
      _highlightResult: {
        title: {
          value: '',
          matchLevel: 'none',
          matchedWords: [],
        },
        url: {
          value: '',
          matchLevel: 'none',
          matchedWords: [],
        },
        author: {
          value: 'xijuan',
          matchLevel: 'none',
          matchedWords: [],
        },
        story_text: {
          value: '26 to 30',
          matchLevel: 'none',
          matchedWords: [],
        },
      },
    },
  ],
  nbHits: 21726300,
  page: 1,
  nbPages: 34,
  hitsPerPage: 30,
  exhaustiveNbHits: true,
  query: '""',
  params:
    'advancedSyntax=true&analytics=true&analyticsTags=backend&hitsPerPage=30&page=1&query=%22%22',
  processingTimeMS: 2,
};

const pushFn = jest.fn();

const props = {
  match: {
    params: { page: null },
  },
  history: {
    push: pushFn,
  },
};

describe('test HomePage', () => {
  const storageSpy = jest.spyOn(Storage.prototype, 'setItem');
  afterEach(() => {
    jest.clearAllMocks();
    storageSpy.mockClear();
  });

  test('should load correct helmet information', () => {
    render(<HomePage {...props} />);
    const helmet = Helmet.peek();
    expect(helmet.title).toStrictEqual('Search Heacker News');
  });

  test('should render loading component on api call', async () => {
    axios.get.mockResolvedValueOnce({
      data: {
        hits: [],
      },
    });

    render(<HomePage {...props} />);

    await waitFor(() => expect(screen.getByTestId('h2loading')).toHaveTextContent('Loading....'));
  });

  test('should render no records component when no records found', async () => {
    axios.get.mockResolvedValueOnce({
      data: {
        hits: [],
      },
    });

    render(<HomePage {...props} />);

    await waitFor(() =>
      expect(screen.getByTestId('norecords')).toHaveTextContent('No Records Available'),
    );
  });

  test('should call api and render data', async () => {
    axios.get.mockResolvedValueOnce({
      data,
    });
    const { asFragment } = render(<HomePage {...props} />);

    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render error component on api fail', async () => {
    axios.get.mockRejectedValueOnce(new Error('API Fail...'));

    render(<HomePage {...props} />);

    await waitFor(() => expect(screen.getByTestId('h2error')).toHaveTextContent('API Fail...'));
  });

  test('should not render previous button when first page loaded', async () => {
    axios.get.mockResolvedValueOnce({
      data,
    });
    render(<HomePage {...props} />);

    await screen.findByTestId('container');

    expect(screen.queryByTestId('previousbtn')).toBeNull();
  });

  test('should redirect to next page on click of next button', async () => {
    axios.get.mockResolvedValueOnce({
      data,
    });
    render(<HomePage {...props} />);

    await screen.findByTestId('container');

    const nextBtn = screen.queryByTestId('nextbtn');

    expect(nextBtn).toBeTruthy();

    fireEvent.click(nextBtn);

    expect(pushFn).toHaveBeenCalledTimes(1);
    expect(pushFn).toHaveBeenCalledWith(`/2`);
  });

  test('should redirect to next page on click of previous button', async () => {
    axios.get.mockResolvedValueOnce({
      data,
    });
    render(
      <HomePage
        {...props}
        match={{
          params: { page: 2 },
        }}
      />,
    );

    await screen.findByTestId('container');

    const previousBtn = screen.queryByTestId('previousbtn');

    expect(previousBtn).toBeTruthy();

    fireEvent.click(previousBtn);

    expect(pushFn).toHaveBeenCalledTimes(1);
    expect(pushFn).toHaveBeenCalledWith(`/`);
  });

  test('should increase upvote count on Upvote click', async () => {
    axios.get.mockResolvedValueOnce({
      data,
    });

    render(<HomePage {...props} />);

    await screen.findByTestId('container');

    const upvote = screen.getByTestId(`upvote-${data.hits[0].objectID}`);

    let currentPoints = data.hits[0].points;

    expect(screen.getByTestId(`td-voteCount-${data.hits[0].objectID}`)).toHaveTextContent(
      currentPoints,
    );

    fireEvent.click(upvote);

    currentPoints += 1;

    await waitFor(() => expect(storageSpy).toHaveBeenCalledTimes(1));
    expect(storageSpy).toHaveBeenCalledWith(
      'upvotes',
      JSON.stringify([{ objectID: data.hits[0].objectID, points: currentPoints }]),
    );

    expect(screen.getByTestId(`td-voteCount-${data.hits[0].objectID}`)).toHaveTextContent(
      currentPoints,
    );

    fireEvent.keyDown(upvote);

    currentPoints += 1;

    await waitFor(() => expect(storageSpy).toHaveBeenCalledTimes(2));
    expect(storageSpy).toHaveBeenCalledWith(
      'upvotes',
      JSON.stringify([{ objectID: data.hits[0].objectID, points: currentPoints }]),
    );

    expect(screen.getByTestId(`td-voteCount-${data.hits[0].objectID}`)).toHaveTextContent(
      currentPoints,
    );
  });

  test('should hide news on click of hide button', async () => {
    axios.get.mockResolvedValueOnce({
      data,
    });

    render(<HomePage {...props} />);

    await screen.findByTestId('container');
    const hideBtn = screen.getByTestId(`hideBtn-${data.hits[0].objectID}`);

    fireEvent.click(hideBtn);

    await waitFor(() => expect(storageSpy).toHaveBeenCalledTimes(1));
    expect(storageSpy).toHaveBeenCalledWith(
      'hiddenHits',
      JSON.stringify([{ objectID: data.hits[0].objectID, hide: true }]),
    );

    expect(screen.queryByTestId(`hideBtn-${data.hits[0].objectID}`)).toBeNull();

    const hideBtn1 = screen.getByTestId(`hideBtn-${data.hits[1].objectID}`);

    fireEvent.keyDown(hideBtn1);

    await waitFor(() => expect(storageSpy).toHaveBeenCalledTimes(2));
    expect(storageSpy).toHaveBeenCalledWith(
      'hiddenHits',
      JSON.stringify([
        { objectID: data.hits[0].objectID, hide: true },
        { objectID: data.hits[1].objectID, hide: true },
      ]),
    );

    expect(screen.queryByTestId(`hideBtn-${data.hits[1].objectID}`)).toBeNull();
  });
});
