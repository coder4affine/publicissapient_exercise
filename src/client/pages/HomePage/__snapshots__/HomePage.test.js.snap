// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test HomePage should call api and render data 1`] = `
<DocumentFragment>
  <div
    data-testid="container"
    style="background: rgb(245, 245, 239);"
  >
    <table
      class="responsive-table"
    >
      <thead>
        <tr>
          <th
            data-testid="th-comments"
            scope="col"
            style="width: 5%;"
          >
            Comments
          </th>
          <th
            data-testid="th-voteCount"
            scope="col"
            style="width: 5%;"
          >
            Vote Count
          </th>
          <th
            data-testid="th-upVote"
            scope="col"
            style="width: 5%;"
          >
            UpVote
          </th>
          <th
            data-testid="th-newsDetails"
            scope="col"
            style="width: 85%; text-align: left;"
          >
            News Details
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-testid="tr-22310813"
        >
          <td
            data-label="Comments"
            data-testid="td-comments-22310813"
          >
            222
          </td>
          <td
            data-label="Vote Count"
            data-testid="td-voteCount-22310813"
          >
            2760
          </td>
          <td
            data-label="UpVote"
            data-testid="td-upVote-22310813"
          >
            <div
              aria-label="upVote for 22310813"
              class="btn"
              data-testid="upvote-22310813"
              role="button"
              tabindex="0"
            >
              <arrow_drop_up.svg
                fill="#979797"
                height="32"
                width="32"
              />
            </div>
          </td>
          <td
            data-label="News Details"
            data-testid="td-newsDetails-22310813"
            style="text-align: left;"
          >
            <div
              style="display: inline-block;"
            >
              <span
                class="mr8 title"
              >
                Gears
              </span>
              <a
                class="mr8"
                href="https://ciechanow.ski/gears/"
                rel="noreferrer"
                target="_blank"
              >
                (https://ciechanow.ski/gears/)
              </a>
              <span
                class="mr8 fs1 clrGray"
              >
                by
              </span>
              <span
                class="mr8 fs1"
              >
                robert-boehnke
              </span>
              <span
                class="mr8 fs1 clrGray"
              >
                4 months ago
              </span>
              <span
                aria-label="hide hit of 22310813"
                class="btn"
                data-testid="hideBtn-22310813"
                role="button"
                tabindex="0"
              >
                [ hide ]
              </span>
            </div>
          </td>
        </tr>
        <tr
          data-testid="tr-5536739"
        >
          <td
            data-label="Comments"
            data-testid="td-comments-5536739"
          >
            0
          </td>
          <td
            data-label="Vote Count"
            data-testid="td-voteCount-5536739"
          >
            2751
          </td>
          <td
            data-label="UpVote"
            data-testid="td-upVote-5536739"
          >
            <div
              aria-label="upVote for 5536739"
              class="btn"
              data-testid="upvote-5536739"
              role="button"
              tabindex="0"
            >
              <arrow_drop_up.svg
                fill="#979797"
                height="32"
                width="32"
              />
            </div>
          </td>
          <td
            data-label="News Details"
            data-testid="td-newsDetails-5536739"
            style="text-align: left;"
          >
            <div
              style="display: inline-block;"
            >
              <span
                class="mr8 title"
              />
              <span
                class="mr8 fs1 clrGray"
              >
                by
              </span>
              <span
                class="mr8 fs1"
              >
                xijuan
              </span>
              <span
                class="mr8 fs1 clrGray"
              >
                about 7 years ago
              </span>
              <span
                aria-label="hide hit of 5536739"
                class="btn"
                data-testid="hideBtn-5536739"
                role="button"
                tabindex="0"
              >
                [ hide ]
              </span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      style="display: flex; flex-direction: row; justify-content: flex-end; color: rgb(191, 72, 8); font-weight: bold; font-size: 1rem; padding: 10px;"
    >
      <span
        aria-label="Next"
        class="btn"
        data-testid="nextbtn"
        role="button"
        tabindex="0"
      >
        Next
      </span>
    </div>
    <div
      style="height: 4px; background-color: rgb(236, 112, 45);"
    />
    <div
      style="margin: 10px;"
    >
      <div
        class="recharts-responsive-container"
        style="width: 100%; height: 300px;"
      >
        <div
          style="position: absolute; width: 0px; height: 0px; visibility: hidden; display: none;"
        />
      </div>
    </div>
    <div
      style="height: 4px; background-color: rgb(236, 112, 45);"
    />
  </div>
</DocumentFragment>
`;
